### PR TITLE
Pagar con Balance

### PR DESCRIPTION
**Pull request by alvvazort**

### CODE REVIEW

File: bugalink-backend/bugalink_backend/settings.py
Comment (pedparbas) [inline bugalink-backend/bugalink_backend/settings.py:1]: Es necesario adjuntar por el canal de discord la modificación sobre el .env ya que desde el settings.py se intenta acceder al token de stripe y no existe. Falla.

File: bugalink-backend/trips/tests.py [hunk 1]
```diff
@@ -59,4 +57,60 @@ def test_get_trip_by_search(self):
...
+            DriverRating.objects.get(trip_request=self.trip_request).rating, 2.3
+        )
+
+class RequestTrip(TestCase):
+    def setUp(self):
+        self.client = APIClient()
+        load_complex_data(self)
+        self.balance = Balance.objects.create(user=self.user_2, amount=100)
+        self.client.force_authenticate(user=self.user_2)
+    '''
```
Comment (pedparbas) [inline bugalink-backend/trips/tests.py:92]: Este comentario es muy grande, ¿es útil para un desarrollo próximo o ya no sirve?

Comment (pedparbas) [inline bugalink-backend/trips/tests.py:92]: Este comentario es muy grande, ¿es útil para un desarrollo próximo o ya no sirve?

File: bugalink-backend/trips/views.py [hunk 2]
```diff
@@ -88,79 +91,105 @@ def get(self, request, *args, **kwargs):
...
+                # Confirm the PaymentMethod to complete the payment
+                payment_intent = stripe.PaymentIntent.create(
+                    payment_method=payment_method.id,
                     amount=amount,
-                    currency="eur",
-                    source=request.data.get(
-                        "stripe_token"
-                    ),  # TODO replace with a valid card token
-                    description="Viaje Bugalink - " + str(trip.departure_datetime),
+                    currency="usd",
```
Comment (pedparbas) [inline bugalink-backend/trips/views.py:121]: EUR no USD

File: bugalink-backend/trips/views.py [hunk 3]
```diff
@@ -253,3 +286,40 @@ def get(self, request, *args, **kwargs):
         return Response(
             TripSerializer(trips, many=True).data, status=status.HTTP_200_OK
         )
+
+
+class TripRateViewSet(
```
Comment (pedparbas) [inline bugalink-backend/trips/views.py:291]: No debería ir en esta PR

File: bugalink-frontend/pages/trips/[id]/pay.tsx [hunk 4]
```diff
@@ -5,8 +5,48 @@ import VisaMastercard from '/public/assets/visa-mastercard.svg';
...
+  if (isLoadingBalance) return <p>Loading...</p>;
+  if (isErrorBalance) return <p>Error</p>;
+
+  const payWithBalance = (event: React.MouseEvent<HTMLButtonElement>) => {
+    if (!isPaying) {
+      event.preventDefault();
+      setIsPaying(true);
+      const data = {
+        payment_method: "Balance",
+        note: "Example note"
```
Comment (pedparbas) [inline bugalink-frontend/pages/trips/[id]/pay.tsx:32]: Example note??

File: bugalink-frontend/pages/trips/[id]/pay.tsx [hunk 5]
```diff
@@ -28,20 +68,24 @@ export default function Pay() {
...
-              data="13,30€"
+              data={`${balance.amount}€`}
               href="#"
+              onClick={payWithBalance}
+              disabled={isPaying}
             />
             <PayMethod
               logo={<VisaMastercard height="100%" />}
               name="VISA/Mastercard"
               data="**** **** **** 5678"
```
Comment (pedparbas) [inline bugalink-frontend/pages/trips/[id]/pay.tsx:79]: Datos genéricos que deberíamos poder ya hacer el retrieve de la base de datos.

File: bugalink-frontend/pages/trips/[id]/pay.tsx [hunk 6]
```diff
@@ -28,20 +68,24 @@ export default function Pay() {
...
               href="#"
+              disabled={isPaying}
             />
             <PayMethod
               logo={<Paypal height="100%" />}
               name="Paypal"
               data="pedro@gmail.com"
```
Comment (pedparbas) [inline bugalink-frontend/pages/trips/[id]/pay.tsx:86]: Datos genéricos...

**Review by pedparbas — CHANGES_REQUESTED** [review]
Comentado todo lo referente al código, las transacciones parecen hacerse correctamente. 

Quedan algunos datos genéricos que deberían haber ya funciones que puedan sustituirlos.

Hay unos cambios que no deberían pertenecer a esta PR (Referido a TripRateViewSet). Borrar en caso de ser así.

Queda a la espera de la limpieza y conexión total con el frontend de momento.

File: bugalink-frontend/pages/trips/[id]/pay.tsx (hunk 5 above)
Comment (andrsdt) [reply bugalink-frontend/pages/trips/[id]/pay.tsx:79]: En cuanto al `**** **** **** 5678`, podemos traernos los 4 últimos dígitos de la CC de la base de datos y en el front lo pintamos con los asteriscos

File: bugalink-backend/bugalink_backend/settings.py [hunk 7]
```diff
@@ -15,6 +15,7 @@
 # SECURITY WARNING: don't run with debug turned on in production!
 DEBUG = config("DEBUG", default=False, cast=bool)
 
+CSRF_TRUSTED_ORIGINS = ["https://app.bugalink.es"]
```
Comment (andrsdt) [inline bugalink-backend/bugalink_backend/settings.py:18]: También se usa `www.app.bugalink.es`

File: bugalink-backend/trips/tests.py [hunk 8]
```diff
@@ -59,4 +57,59 @@ def test_get_trip_by_search(self):
...
+                "is_pleasant_driver": False,
+                "already_knew": True,
+            },
+        )
+
+        self.assertEqual(
+            DriverRating.objects.get(trip_request=self.trip_request).rating, 2.3
+        )
+
+class RequestTrip(TestCase):
```
Comment (andrsdt) [inline bugalink-backend/trips/tests.py:86]: These are tests "pa salir del paso" as they don't test any unitary functionality, if-else branches... such that the method could not work as expected and still return a code 201. In fact, there aren't any tests that make sure that invalid data results in errors and the objects not being created in the DB, which are probably the most important tests (everyone tests the happy path but there are always edge cases and misuse in the app)

File: bugalink-backend/trips/views.py [hunk 9]
```diff
@@ -88,79 +94,124 @@ def get(self, request, *args, **kwargs):
...
+            })
+
+            # Create a payment object
+            payment = Payment({
+                "intent": "sale",
+                "payer": {
+                    "payment_method": "paypal",
+                },
+                "redirect_urls": {
+                    "return_url": "http://app.bugalink.es",
```
Comment (andrsdt) [inline bugalink-backend/trips/views.py:161]: What about development?? This has to use an environment variable defining that the front URL is either this or `localhost:3000`. Also, you are missing the "s" in "https://"

File: bugalink-frontend/pages/trips/[id]/pay.tsx [hunk 10]
```diff
@@ -28,20 +68,24 @@ export default function Pay() {
             <PayMethod
               logo={<BugalinkLogo color="white" />}
               name="Saldo"
-              data="13,30€"
+              data={`${balance.amount}€`}
```
Comment (andrsdt) [inline bugalink-frontend/pages/trips/[id]/pay.tsx:71]: use `.toLocaleString()` as in other pages where we format numbers into money

**Review by andrsdt — CHANGES_REQUESTED** [review]
There are some important details to change before merging, apart from verifying that the functionality works, of course

File: bugalink-backend/trips/views.py (hunk 3 above)
Comment (alvvazort) [reply bugalink-backend/trips/views.py]: Deleted

File: bugalink-frontend/pages/trips/[id]/pay.tsx (hunk 4 above)
Comment (alvvazort) [reply bugalink-frontend/pages/trips/[id]/pay.tsx]: Now is a real note from the user

File: bugalink-frontend/components/payment/CreditOperation.tsx [hunk 11]
```diff
@@ -29,17 +29,24 @@ export default function CreditOperation({
               logo={<VisaMastercard />}
               name="VISA/Mastercard"
               data="**** **** **** 5678"
-              href="#"
+              onClick={() => {
+                return;
```
Comment (pedparbas) [inline bugalink-frontend/components/payment/CreditOperation.tsx:33]: Dejar al menos un TO-DO para que quede reflejado y se vea más fácilmente

File: bugalink-frontend/components/payment/CreditOperation.tsx [hunk 12]
```diff
@@ -29,17 +29,24 @@ export default function CreditOperation({
...
+              }}
             />
             <PayMethod
               logo={<Paypal />}
               name="Paypal"
               data="pedro@gmail.com"
-              href="#"
+              onClick={() => {
+                return;
```
Comment (pedparbas) [inline bugalink-frontend/components/payment/CreditOperation.tsx:41]: Lo mismo que en el comentario anterior

**Review by pedparbas — APPROVED** [review]
Funciona 👍🏾
